File: ts/src/redact.ts
/**
 * Argument redaction and canonical digests.
 *
 * Redaction is ON by default: the inline copy of tool/prompt arguments has
 * sensitive values replaced with "[REDACTED]" before it ever reaches an
 * exporter. A SHA-256 digest of the canonical JSON encoding of the ORIGINAL
 * arguments is always recorded, so two records can still be correlated (and
 * tampering detected) without storing the sensitive payload.
 */
import { createHash } from "node:crypto";
import type { ArgumentsDigest } from "./types.js";

/** Replacement string used for redacted values. */
export const REDACTED = "[REDACTED]";

/**
 * Key-name deny list. A key matches when its normalized form (lowercase,
 * `-`/`_` removed) contains one of these substrings.
 */
export const DEFAULT_SENSITIVE_KEY_PARTS: readonly string[] = [
  "password",
  "passwd",
  "secret",
  "token",
  "apikey",
  "authorization",
  "credential",
  "privatekey",
  "accesskey",
  "sessionkey",
  "cookie",
];

/** Keys that match exactly (after normalization) even though they are short. */
export const DEFAULT_SENSITIVE_KEY_EXACT: readonly string[] = ["auth", "ssn", "pin", "otp"];

/**
 * Value patterns treated as secrets regardless of the key they appear under:
 * JWTs, AWS access key ids, GitHub tokens, OpenAI-style keys, Bearer values.
 */
export const DEFAULT_SENSITIVE_VALUE_PATTERNS: readonly RegExp[] = [
  /^eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{5,}\.[A-Za-z0-9_-]{5,}$/, // JWT
  /\bAKIA[0-9A-Z]{16}\b/, // AWS access key id
  /\bgh[pousr]_[A-Za-z0-9]{20,}\b/, // GitHub token
  /\bsk-[A-Za-z0-9_-]{20,}\b/, // OpenAI-style secret key
  /^Bearer\s+\S+$/i, // Authorization: Bearer <...>
];

/** Redaction behaviour for the inline arguments copy. */
export type RedactionMode = "omit" | "redact" | "raw";

/** Redaction configuration accepted by AuditLogger and withAudit. */
export interface RedactionPolicy {
  /**
   * omit   – no inline copy at all, digest only.
   * redact – inline copy with sensitive values replaced (default).
   * raw    – inline copy unchanged. Only for closed environments.
   */
  mode?: RedactionMode;
  /** Extra key substrings to treat as sensitive (normalized matching). */
  extraSensitiveKeys?: string[];
  /** Extra value patterns to treat as sensitive. */
  extraValuePatterns?: RegExp[];
}

export interface RedactionResult {
  /** Deep copy of the arguments with sensitive values replaced. */
  redacted: Record<string, unknown> | null;
  /** Dot-paths of the keys that were redacted. */
  redactedKeys: string[];
}

function normalizeKey(key: string): string {
  return key.toLowerCase().replace(/[-_]/g, "");
}

function isSensitiveKey(key: string, extraParts: string[]): boolean {
  const norm = normalizeKey(key);
  if (DEFAULT_SENSITIVE_KEY_EXACT.includes(norm)) return true;
  for (const part of DEFAULT_SENSITIVE_KEY_PARTS) {
    if (norm.includes(part)) return true;
  }
  for (const part of extraParts) {
    if (norm.includes(normalizeKey(part))) return true;
  }
  return false;
}

function isSensitiveValue(value: string, extraPatterns: RegExp[]): boolean {
  for (const re of DEFAULT_SENSITIVE_VALUE_PATTERNS) {
    if (re.test(value)) return true;
  }
  for (const re of extraPatterns) {
    if (re.test(value)) return true;
  }
  return false;
}

function redactValue(
  value: unknown,
  path: string,
  policy: Required<Pick<RedactionPolicy, "extraSensitiveKeys" | "extraValuePatterns">>,
  redactedKeys: string[]
): unknown {
  if (typeof value === "string") {
    if (isSensitiveValue(value, policy.extraValuePatterns)) {
      redactedKeys.push(path);
      return REDACTED;
    }
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((item, i) => redactValue(item, `${path}[${i}]`, policy, redactedKeys));
  }
  if (value !== null && typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const [key, child] of Object.entries(value as Record<string, unknown>)) {
      const childPath = path === "" ? key : `${path}.${key}`;
      if (isSensitiveKey(key, policy.extraSensitiveKeys) && child !== null && child !== undefined) {
        redactedKeys.push(childPath);
        out[key] = REDACTED;
      } else {
        out[key] = redactValue(child, childPath, policy, redactedKeys);
      }
    }
    return out;
  }
  return value;
}

/** Apply a redaction policy to a set of arguments. */
export function redactArguments(
  args: Record<string, unknown> | null | undefined,
  policy: RedactionPolicy = {}
): RedactionResult {
  if (args === null || args === undefined) {
    return { redacted: null, redactedKeys: [] };
  }
  const redactedKeys: string[] = [];
  const redacted = redactValue(
    args,
    "",
    {
      extraSensitiveKeys: policy.extraSensitiveKeys ?? [],
      extraValuePatterns: policy.extraValuePatterns ?? [],
    },
    redactedKeys
  ) as Record<string, unknown>;
  return { redacted, redactedKeys };
}

/**
 * Canonical JSON encoding: object keys sorted lexicographically at every
 * level, no insignificant whitespace, UTF-8. Matches the Python SDK's
 * `json.dumps(obj, sort_keys=True, separators=(",", ":"))` for the JSON
 * subset both languages produce.
 */
export function canonicalJson(value: unknown): string {
  return JSON.stringify(sortValue(value));
}

function sortValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortValue);
  if (value !== null && typeof value === "object") {
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      sorted[key] = sortValue((value as Record<string, unknown>)[key]);
    }
    return sorted;
  }
  return value;
}

/** Compute the integrity digest of the original (pre-redaction) arguments. */
export function digestArguments(
  args: Record<string, unknown> | null | undefined,
  redactedKeys: string[] = []
): ArgumentsDigest {
  const canonical = canonicalJson(args ?? null);
  const bytes = Buffer.from(canonical, "utf8");
  const digest: ArgumentsDigest = {
    sha256: createHash("sha256").update(bytes).digest("hex"),
    byte_length: bytes.byteLength,
  };
  if (redactedKeys.length > 0) digest.redacted_keys = redactedKeys;
  return digest;
}

File: python/mcp_audit/redact.py
"""Argument redaction and canonical digests.

Redaction is ON by default: the inline copy of tool/prompt arguments has
sensitive values replaced with "[REDACTED]" before it ever reaches an
exporter. A SHA-256 digest of the canonical JSON encoding of the ORIGINAL
arguments is always recorded, so two records can still be correlated (and
tampering detected) without storing the sensitive payload.
"""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Pattern, Tuple

REDACTED = "[REDACTED]"

# Key-name deny list. A key matches when its normalized form (lowercase,
# "-"/"_" removed) contains one of these substrings.
DEFAULT_SENSITIVE_KEY_PARTS: Tuple[str, ...] = (
    "password",
    "passwd",
    "secret",
    "token",
    "apikey",
    "authorization",
    "credential",
    "privatekey",
    "accesskey",
    "sessionkey",
    "cookie",
)

# Keys that match exactly (after normalization) even though they are short.
DEFAULT_SENSITIVE_KEY_EXACT: Tuple[str, ...] = ("auth", "ssn", "pin", "otp")

# Value patterns treated as secrets regardless of the key they appear under:
# JWTs, AWS access key ids, GitHub tokens, OpenAI-style keys, Bearer values.
DEFAULT_SENSITIVE_VALUE_PATTERNS: Tuple[Pattern[str], ...] = (
    re.compile(r"^eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{5,}\.[A-Za-z0-9_-]{5,}$"),
    re.compile(r"\bAKIA[0-9A-Z]{16}\b"),
    re.compile(r"\bgh[pousr]_[A-Za-z0-9]{20,}\b"),
    re.compile(r"\bsk-[A-Za-z0-9_-]{20,}\b"),
    re.compile(r"^Bearer\s+\S+$", re.IGNORECASE),
)


@dataclass
class RedactionPolicy:
    """Redaction configuration accepted by AuditLogger.

    mode:
      omit   -- no inline copy at all, digest only.
      redact -- inline copy with sensitive values replaced (default).
      raw    -- inline copy unchanged. Only for closed environments.
    """

    mode: str = "redact"
    extra_sensitive_keys: List[str] = field(default_factory=list)
    extra_value_patterns: List[Pattern[str]] = field(default_factory=list)

    def __post_init__(self) -> None:
        if self.mode not in ("omit", "redact", "raw"):
            raise ValueError(f"invalid redaction mode: {self.mode!r}")


def _normalize_key(key: str) -> str:
    return key.lower().replace("-", "").replace("_", "")


def _is_sensitive_key(key: str, extra_parts: List[str]) -> bool:
    norm = _normalize_key(key)
    if norm in DEFAULT_SENSITIVE_KEY_EXACT:
        return True
    for part in DEFAULT_SENSITIVE_KEY_PARTS:
        if part in norm:
            return True
    for part in extra_parts:
        if _normalize_key(part) in norm:
            return True
    return False


def _is_sensitive_value(value: str, extra_patterns: List[Pattern[str]]) -> bool:
    for pattern in DEFAULT_SENSITIVE_VALUE_PATTERNS:
        if pattern.search(value):
            return True
    for pattern in extra_patterns:
        if pattern.search(value):
            return True
    return False


def _redact_value(value: Any, path: str, policy: RedactionPolicy, redacted_keys: List[str]) -> Any:
    if isinstance(value, str):
        if _is_sensitive_value(value, policy.extra_value_patterns):
            redacted_keys.append(path)
            return REDACTED
        return value
    if isinstance(value, list):
        return [
            _redact_value(item, f"{path}[{i}]", policy, redacted_keys)
            for i, item in enumerate(value)
        ]
    if isinstance(value, dict):
        out: Dict[str, Any] = {}
        for key, child in value.items():
            child_path = key if path == "" else f"{path}.{key}"
            if _is_sensitive_key(str(key), policy.extra_sensitive_keys) and child is not None:
                redacted_keys.append(child_path)
                out[key] = REDACTED
            else:
                out[key] = _redact_value(child, child_path, policy, redacted_keys)
        return out
    return value


def redact_arguments(
    args: Optional[Dict[str, Any]], policy: Optional[RedactionPolicy] = None
) -> Tuple[Optional[Dict[str, Any]], List[str]]:
    """Apply a redaction policy; returns (redacted_copy, redacted_key_paths)."""
    if args is None:
        return None, []
    policy = policy or RedactionPolicy()
    redacted_keys: List[str] = []
    redacted = _redact_value(args, "", policy, redacted_keys)
    return redacted, redacted_keys


def canonical_json(value: Any) -> str:
    """Canonical JSON: sorted keys, no insignificant whitespace, UTF-8.

    Matches the TypeScript SDK's canonicalJson for the JSON subset both
    languages produce.
    """
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def digest_arguments(
    args: Optional[Dict[str, Any]], redacted_keys: Optional[List[str]] = None
) -> Dict[str, Any]:
    """Compute the integrity digest of the original (pre-redaction) arguments."""
    canonical = canonical_json(args if args is not None else None)
    raw = canonical.encode("utf-8")
    digest: Dict[str, Any] = {
        "sha256": hashlib.sha256(raw).hexdigest(),
        "byte_length": len(raw),
    }
    if redacted_keys:
        digest["redacted_keys"] = list(redacted_keys)
    return digest
